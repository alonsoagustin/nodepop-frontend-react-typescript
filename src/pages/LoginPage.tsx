import { useState } from "react";
import { useNavigate, useLocation, Link } from "react-router-dom";
import { login } from "../auth/service";
import { useAuth } from "../auth/context";
import Button from "../components/Button";

const LoginPage = () => {
  // Hook to manage the authentication state
  const { onLogin } = useAuth();

  // Hook to manage the navigation
  const navigate = useNavigate();

  // Hook to manage the location
  const location = useLocation();

  // Hook to manage the inputs state
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [checked, setChecked] = useState(true);

  const isDisabled = !email || !password;

  const handleEmailChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setEmail(e.target.value);
  };

  const handlePasswordChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setPassword(e.target.value);
  };

  const handleRememberChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setChecked(e.target.checked);
  };

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    try {
      const credentials = { email, password };
      const response = await login(credentials);

      if (checked) {
        localStorage.setItem("accessToken", response);
      }

      // Update the authentication state
      onLogin();

      // Redirect to the previous page or to the home page
      navigate(location.state?.from ?? "/", { replace: true });
    } catch (error) {
      console.error("Error while trying to login", error);
    }
  };

  return (
    <>
      <h2 className="loginPage-title">ACCESS YOUR ACCOUNT</h2>
      <form className="loginPage-form" onSubmit={handleSubmit}>
        <div className="formField">
          <label className="formField-label" htmlFor="email">
            <span>Email:</span>
          </label>
          <input
            className="formField-input"
            type="email"
            name="email"
            id="email"
            placeholder="Insert your email here"
            autoComplete="on"
            value={email}
            onChange={handleEmailChange}
            required
          />
        </div>
        <div className="formField">
          <label className="formField-label" htmlFor="password">
            <span>Password:</span>
          </label>
          <input
            className="formField-input"
            type="password"
            name="password"
            id="password"
            placeholder="Insert your password here"
            value={password}
            onChange={handlePasswordChange}
            required
          />
        </div>
        <div className="formField">
          <label className="formField-label" htmlFor="remember">
            <span>Remember me</span>
          </label>
          <input
            className="formField-input"
            type="checkbox"
            name="remember"
            id="remember"
            checked={checked}
            onChange={handleRememberChange}
          />
        </div>
        <div className="formField">
          <Link to={"/signup"}>
            <span>I don't have an account yet</span>
          </Link>
        </div>
        <div className="formActions">
          <Button
            className={"loginForm-submit btn"}
            type={"submit"}
            disabled={isDisabled}
          >
            Login
          </Button>
        </div>
      </form>
    </>
  );
};

export default LoginPage;
