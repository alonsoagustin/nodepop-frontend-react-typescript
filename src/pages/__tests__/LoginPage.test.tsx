import { render, screen } from "@testing-library/react";
import LoginPage from "../LoginPage";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";

const state = {
  auth: false,
  ui: { error: "", loading: false },
};

const renderComponent = () =>
  render(
    <BrowserRouter>
      <Provider
        store={{
          getState: () => state,
          subscribe: () => {},
          dispatch: () => {},
        }}
      >
        <LoginPage />
      </Provider>
    </BrowserRouter>
  );

describe("LoginPage", () => {
  test("should render login form", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });
});
