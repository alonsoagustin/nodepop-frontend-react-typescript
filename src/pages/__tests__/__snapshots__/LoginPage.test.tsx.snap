// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginPage > should render login form 1`] = `
<div>
  <h2
    class="text-center mb-4"
  >
    ACCESS YOUR ACCOUNT
  </h2>
  <div
    class="container"
  >
    <div
      class="row d-flex justify-content-center"
    >
      <form
        class="col-12 col-md-6 col-lg-4 p-4 border rounded"
      >
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="email"
          >
            <span>
              Email adress
            </span>
             
          </label>
          <input
            class="form-control"
            id="email"
            placeholder="Insert your email here"
            required=""
            type="email"
            value=""
          />
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="password"
          >
            <span>
              Password
            </span>
             
          </label>
          <input
            class="form-control"
            id="password"
            placeholder="Insert your password here"
            required=""
            type="password"
            value=""
          />
        </div>
        <div
          class="form-check mb-0"
        >
          <input
            checked=""
            class="form-check-input"
            id="remember"
            name="remember"
            type="checkbox"
            value="remember"
          />
          <label
            class="form-label"
            for="remember"
          >
            <span>
              Remember me
            </span>
          </label>
        </div>
        <div
          class="mb-3"
        >
          <a
            data-discover="true"
            href="/signup"
          >
            <span
              class="btn btn-link"
            >
              I don't have an account yet
            </span>
          </a>
        </div>
        <div
          class="d-flex justify-content-end "
        >
          <button
            class="btn btn btn-primary"
            disabled=""
            type="submit"
          >
            Login
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
